### PR TITLE
READY: Session refactor

### PR DESCRIPTION
**Pull request by ichorid**
* Remove LaunchMany and redistribute its functionality between Session and LibtorrentMgr
 * Move all torrent-related methods to LibtorrentMgr

### CODE REVIEW

**Comment by ichorid** [comment]
retest this please

File: Tribler/Core/Libtorrent/LibtorrentMgr.py [hunk 1]
```diff
@@ -631,9 +655,326 @@ def post_session_stats(self, hops=None):
...
+
+        config = config or self.load_download_config_by_infohash(infohash)  # not already resuming
+
+        # Store in list of Downloads, always.
+        self.downloads[infohash] = download
+        download.setup(config, delay=delay, checkpoint_disabled=checkpoint_disabled,
+                       hidden=hidden or config.get_bootstrap_download())
+        return download
+
+    async def remove(self, download, removecontent=False, removestate=True):
```
Comment (egbertbouman) [inline Tribler/Core/Libtorrent/LibtorrentMgr.py]: Why rename these keyword arguments. What's wrong with using underscores to separate words?

File: Tribler/Core/Libtorrent/LibtorrentMgr.py [hunk 2]
```diff
@@ -59,6 +80,9 @@ def __init__(self, tribler_session):
 
         self.torrents = {}
 
+        # TODO: merge with self.torrents. Legacy stuff moved from Session
```
Comment (egbertbouman) [inline Tribler/Core/Libtorrent/LibtorrentMgr.py]: Can't we just do this now? Seems pretty straightforward.

File: Tribler/Core/Libtorrent/LibtorrentMgr.py [hunk 3]
```diff
@@ -631,9 +655,326 @@ def post_session_stats(self, hops=None):
         elif hasattr(self.ltsessions[hops], "post_session_stats"):
             self.ltsessions[hops].post_session_stats()
 
+    #####################################
+    # MOVED FROM LAUNCHMANY AND SESSION #
```
Comment (egbertbouman) [inline Tribler/Core/Libtorrent/LibtorrentMgr.py]: Not sure why this is here. We can probably do without this.

File: Tribler/Core/Libtorrent/LibtorrentMgr.py [hunk 4]
```diff
@@ -631,9 +655,326 @@ def post_session_stats(self, hops=None):
...
+                        metainfo["announce-list"] = [all_trackers]
+                    else:
+                        metainfo["announce"] = all_trackers[0]
+                    new_def = TorrentDef.load_from_dict(metainfo)
+
+                # Set TorrentDef + checkpoint
+                dl.set_def(new_def)
+                dl.checkpoint()
+
+    # ???????
```
Comment (egbertbouman) [inline Tribler/Core/Libtorrent/LibtorrentMgr.py]: Should this really be in devel?

File: Tribler/Core/CreditMining/CreditMiningManager.py [hunk 5]
```diff
@@ -102,15 +102,15 @@ def add_sources(future):
         self.session_ready = Future()
         self.session_ready.add_done_callback(add_sources)
 
-    async def shutdown(self, remove_downloads=False):
+    async def shutdown(self, removes=False):
```
Comment (egbertbouman) [inline Tribler/Core/CreditMining/CreditMiningManager.py]: Why rename this variable?

File: Tribler/Core/Libtorrent/LibtorrentMgr.py [hunk 6]
```diff
@@ -631,9 +655,326 @@ def post_session_stats(self, hops=None):
...
+
+        if self.state_cb_count % 4 == 0:
+            if self.tribler_session.tunnel_community:
+                self.tribler_session.tunnel_community.monitor_downloads(states_list)
+            if self.tribler_session.credit_mining_manager:
+                self.tribler_session.credit_mining_manager.monitor_downloads(states_list)
+
+        return []
+
+    # TODO: Make it async?
```
Comment (egbertbouman) [inline Tribler/Core/Libtorrent/LibtorrentMgr.py]: This is just adding downloads, which as far as I know is already async (although the function itself is not defined as `async`).

File: Tribler/Core/CreditMining/CreditMiningManager.py (hunk 5 above)
Comment (ichorid) [reply Tribler/Core/CreditMining/CreditMiningManager.py]: Oops. That's one of the consequences of, seemingly, a bit too broad `sed` editing that I had to do on the codebase.

File: Tribler/Core/Libtorrent/LibtorrentMgr.py (hunk 2 above)
Comment (ichorid) [reply Tribler/Core/Libtorrent/LibtorrentMgr.py]: The intention of this PR was to take the basic, most literary steps to remove LaunchMany and move stuff from Session. There are lots of opportunities to optimize stuff further in smaller PRs.

File: Tribler/Core/Libtorrent/LibtorrentMgr.py (hunk 3 above)
Comment (ichorid) [reply Tribler/Core/Libtorrent/LibtorrentMgr.py]: That's to help the reader understand that the following methods were not added here by design, but instead is legacy code. I would prefer it to stay here until the moment when we refactor Libtorren wrapper properly.

File: Tribler/Core/Libtorrent/LibtorrentMgr.py (hunk 1 above)
Comment (ichorid) [reply Tribler/Core/Libtorrent/LibtorrentMgr.py]: Nothing wrong with the names. It was just easier to change `remove_state` for all the calls to `remove_download` method to become the same as `removestate` and `removecontent` arguments used by `metainfo_dl.stop`, than to the reverse.

File: Tribler/Core/Libtorrent/LibtorrentMgr.py (hunk 4 above)
Comment (ichorid) [reply Tribler/Core/Libtorrent/LibtorrentMgr.py]: Oops, that's a leftover, gonna remove it...

File: Tribler/Core/Libtorrent/LibtorrentMgr.py (hunk 6 above)
Comment (ichorid) [reply Tribler/Core/Libtorrent/LibtorrentMgr.py]: Good point, removed notice.

File: Tribler/Core/Libtorrent/LibtorrentMgr.py (hunk 3 above)
Comment (ichorid) [reply Tribler/Core/Libtorrent/LibtorrentMgr.py]: Removed that

File: Tribler/Core/Libtorrent/LibtorrentMgr.py (hunk 1 above)
Comment (ichorid) [reply Tribler/Core/Libtorrent/LibtorrentMgr.py]: Renamed to use underscore

**Pull request merged**